### PR TITLE
Give decoder input an nfeat dimension

### PR DESCRIPTION
**Pull request by bpopeters**
This pull request makes a few small changes to IO.py, Models.py, Translator.py, and train.py. The goal of the changes is to make it to get rid of some of the artificial barriers that are currently standing in the way of implementing target features.

I have eliminated two of these obstacles:

1. Previously, `Decoder.forward()` expected an input whose dimensions were length x batch_size. A third dimension, analogous to the nfeat dimension used in the encoder, is added. Tensor logic is changed as necessary in the decoder to make use of these new dimensions.

2. Code in other files is changed as necessary so that the target variables passed to the decoder will be the correct size. This required abstracting out `make_features` in IO.py: now it accepts a batch object and a side (src or tgt) and constructs the length x batch_size x nfeat tensor. I'm not convinced that IO.py is the right place to be doing tensor operations like that, but it's the status quo. Consequently, `make_features` no longer needs the `ONMTDataset.collect_features` static method, which is hard-coded to only work for src sequences. I'm not certain if that function is necessary anywhere else, so for now I did not delete it. I then needed to call the new feature-making code everywhere where data is passed through the model: in train.py and translator.py.

### CODE REVIEW

File: onmt/IO.py [hunk 1]
```diff
@@ -56,12 +56,20 @@ def merge_vocabs(vocabs, vocab_size=None):
...
 
-def make_features(batch, fields):
-    # TODO: This is bit hacky, add to batch somehow.
-    f = ONMTDataset.collect_features(fields)
-    cat = [batch.src[0]] + [batch.__dict__[k] for k in f]
-    cat = [c.unsqueeze(2) for c in cat]
-    return torch.cat(cat, 2)
+def make_features(batch, side):
+    """
+    data (Variable): a batch of source or target data
```
Comment (srush) [inline onmt/IO.py]: Looks like this is called `batch` now.

**Comment by srush** [comment]
Looks good to me. This asymmetry was bothering me as well. 

Shouldn't be that hard to add tgt features to the loss. Beam search will be a bit annoying though as always.

**Comment by srush** [comment]
btw, we are going to switch to use `pep8-naming` which should help find all the bad `camelCase` issues
